Technical Design documentation (#36)
kyma-project/istio: docs/technical-design.md

which is used to make changes to the Istio installation by passing it to the `IstioManager`.

The installed IstioOperator is created by merging the `IstioCR` with the IstioOperator with Kyma default values.

##### IstioManager
This component contains the logic for managing the Istio installation. It knows about the supported client versions and forwards the 
Istio API requests (e.g. Install, Upgrade) to the correct version of `IstioClient` if there are breaking changes.  
Depending on how the Istio Go module behaves, it may not be necessary to support multiple clients for the time being. For the first simple implementation, we can probably do without the IstioManager for the time being.

##### IstioClient
A IstioClient encapsulates a specific version of the [Istio Go module](https://github.com/istio/istio). 
As we want to support canary updates at some point we might need to support two version of the library if there are breaking changes.

#### ProxySidecarReconciler
This reconciler must be executed after the `IstioInstallationReconciler`. Its responsibility is to restart pods based on specific configuration changes.

> As of now the following scenarios must be covered by this reconciler:

Reviewer: We clarified that the sub-reconcilers will be part of the main reconciler deployment, and they will run as needed

# Technical Design

## To be discussed

## Istio operator

![IstioOperator Overview](./istio-operator-overview.svg)

### Ownership of current resources in Kyma repository

In the transitioning to a more modularised architecture, the [IstioOperator resource](https://github.com/kyma-project/kyma/tree/main/resources/istio), 
the [additional istio-resources](https://github.com/kyma-project/kyma/tree/main/resources/istio-resources) and 
the [certificates](https://github.com/kyma-project/kyma/tree/main/resources/certificates) must be moved to the new modules.

#### IstioOperator resource
This is moved into our new Istio operator and used to define the default values for Istio, which can be customised by the user through `IstioCR`.

#### istio-resources

##### Istio Grafana dashboards
It's still needs to be decided who will have the ownership of the dashboards. There are different things to consider like the change interval or relevance of Istio version updates.

##### Istio ServiceMonitor & istio-healthz Virtual Service

> They provide monitoring capability for the Istio installation and should therefore be reconciled by the operator.

Reviewer 2: I'm planning to replace the direct dependency by a simple annotation: https://github.com/kyma-project/kyma/pull/16247

##### Global mTLS PeerAuthentication
This is tightly coupled to our Istio installation and should therefore be reconciled by the operator.

##### Kyma Gateway
This should be moved to the API Gateway as it's a default gateway we want to provide, and it has a stronger relation to the responsibilities
of the API Gateway that to Istio. Since API Gateway is already dependent on Istio, we do not add any additional dependency by moving it. 

#### Certificate resources
They should also be moved to API Gateway as they are tightly coupled with the `Kyma Gateway` resource.

### Handling of Istio version
We don't want to expose the version. That means that the version of Istio is coupled to the version of the operator. The benefit of this is,
that we are in full control of the versioning and hide this complexity from the user.  
That means if we want to release a new Istio version, we have to release a new version of the operator.

## IstioController

![Component Diagram](./controller-component-diagram.svg)

### Installation & Upgrade of Istio
The Istio installation, upgrade and uninstall is done using [Istio Go module](https://github.com/istio/istio).

### Reconciliation

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Author: I simplified the structure in the component diagram and added some [additional docs](https://github.com/triffer/istio/blob/7ba855023208378cd612d03a9d5782df901dcb7e/docs/technical-design.md?plain=1#L50) to explain our relation between the controller and reconcilers. This means your assumptions that what we called reconcilers are actually controllers doesn't match what we had in mind when creating this modular controller design.
I agree that it might make sense to split our "sub-reconciler" up in multiple controllers, but I see some drawbacks about this. 
Some of the new controllers will use the same CR and rely on fields that are rolled out by the IstioController (e.g. a controller handling the logic of `IstioIngressGatewayReconciler`. Other controllers will be fully time-based and rely on the state of Istio installation, e.g. controller handling the logic of `PeerAuthenticationReconciler`.
Decoupling this from the Istio reconciliation will make the implementation of this controllers more complex and doesn't provide any real benefit in my opinion. Since the logic is already divided into components, it should be easy to move them to separate controllers if needed.

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Author: I added a simple sequence diagram for the reconciliation loop of `IstioController`. You can find it [here](https://github.com/triffer/istio/blob/technical-design/docs/reconciliation-sequence-diagram.svg). 
We will also rename components that had the suffix `Reconciler` to `Reconciliation`.
I hope this helps to better understand the components.

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Reviewer 3: Ok. It is more clear now. Still it is important to avoid sequences and waiting in the reconciliation logic. If you need to invoke second reconciliation (e.g. proxy sidecar) you should not do it in the same main loop invocation as IstioInstallation. You trigger IstioInstallation in the first invocation and you should return RequeAfter. The second invocation should check if installation is completed and execute proxy sidecar reconciliation. I think you know it, but initial description and sequence diagram can be misleading.

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Reviewer 4: Reconcile of a controller is meant to be fast and non-waiting process. Installation is actually not really optimal for such a task. I did express this concern a while back, but somehow did not trigger discussion about it. What do you think of moving it out of the reconcile() processing? For instance into a separate Job task? We used before Argo Workflows for various long running tasks. I would like that the Istio operator only control operations of others than to actually do time consuming tasks.

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Reviewer 3: Installation can be done this way as it is just about creating some kubernetes resources (request to api server) and waiting for the status. You don't need to wait. You should check the status and if it is not ready exit the reconciliation loop with RequeueAfter. In the next invocation you check it again and if nothing has changed you exit it again. You don't have anything to process (any time consuming task).

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Reviewer 4: We install Istio with the CLI which takes time? or I am mistaken something. Current Reconciler run looks like that:
```
{"level":"INFO","time":"2022-12-08T13:36:55.334+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:37:25.348+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:37:38.557+0100","message":"Istio has been updated successfully to version 1.16.0-distroless"}
{"level":"INFO","time":"2022-12-08T13:37:38.557+0100","message":"Restarting ingress-gateway"}
2022-12-08T12:37:42.753965Z	info	klog	Waited for 1.146225584s due to client-side throttling, not priority and fairness, request: PATCH:https://api.vv-test2.goatz.shoot.canary.k8s-hana.ondemand.com/apis/apps/v1/namespaces/vladimir/deployments/httpbin
{"level":"INFO","time":"2022-12-08T13:37:55.350+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:38:25.351+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:38:55.352+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:39:25.352+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:39:33.442+0100","message":"Proxy reset for 17 pods successfully done"}
{"level":"WARN","time":"2022-12-08T13:39:34.250+0100","message":"Reconciler Metrics not initialized"}
{"level":"INFO","time":"2022-12-08T13:39:34.395+0100","message":"Component 'istio' has status 'success'"}
```
Looking at the timestamps there is quite something going on. My idea was to extract these time consuming operations outside of the controller. Having them in separate Jobs it may also be easy on analysing logs, when something goes wrong. Not needed to search huge reconciler logs. Let's discuss it tomorrow.

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Reviewer 3: Timestamp doesn't say that istio installer is doing anything. Istio uses helm to render manifests, then applies the manifest in the kuybernetes and then waits. And waits. And waits. Active part is probably 1%. The rest is waiting for kubelet to deploy istiod and istioingressgateway.

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Reviewer 4: We like the Istio Operator Install, we just are being hesitant because of the `discouraged` note in here:
https://istio.io/latest/docs/setup/install/operator/

> The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility

Author: @pbochynski 
We updated the documentation for the [Istio installation](https://github.com/kyma-project/istio/pull/36/files#diff-60b0177084514f6a20db25c5360a874044da6414583268b928d0056eeceb308eR50). 
We decided to install the operator using the Istio Go module and apply the `IstioOperator CR` to install Istio. This way we avoid the long blocking the reconciliation of our controller during the Istio Installation. We also added a flow diagram to visualise the process of installing, upgrading and uninstalling Istio.

The docs were extended to clarify the scenario when the `IstioVersionManager` (formerly known as `IstioManager`) would be useful.

and must work in isolation when assessing whether reconciliation is required, applying changes and returning a status.  

Although we want the reconcilers to be as decoupled and independent as possible, there is an execution dependency as we first need to install/upgrade Istio ( done by `IstioInstallationReconciler`)
before the other reconcilers can be executed.

#### Interval

Since this module deals with security-related topics, we want to perform the reconciliation as often as possible.

> This means that we do not only want to reconcile when `IstioCR` changes, but also want to check time-based that resources have not been changed and are in the expected state.  

Reviewer 3: I was confused by the name. Add a space between Istio and CR - you meant Istio CR :)

}
```

The biggest challenge in deciding on an appropriate interval is that the time required to perform the reconciliation can vary a lot. Small changes may only require a 
restart of the sidecar proxies or the Ingress gateway and are therefore much faster than a new installation or a Canary upgrade.  
Given this dynamic in execution, we must always consider the slowest reconciliation process when defining `SyncPeriod`. Also, we have to consider that if we later add another controller to this operator, this new controller will also use the same frequency.  
Using `RequeueAfter` on the other hand gives us more freedom, but also makes it a bit more complex to understand what is being done and why.

We have decided to use `RequeueAfter` with a frequency of 5 minutes, as this gives us the ability to perform the matching as often as possible without the risk of repeatedly checking the slowest reconciliation process or piling up requests in the reconciliation queue.

The queuing of reconciliation requests is handled by [controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime) and is out of scope of this design.

### Components
#### IstioReconciler

> This is the reconciler that takes care of the entire Istio reconciliation process. 

Author: i think this might be partly answered by my reply [here](https://github.com/kyma-project/istio/pull/36#discussion_r1042104795). As the desired and actual state should be represented by the [IstioCR](https://github.com/kyma-project/istio/blob/main/docs/xff-proposal.md) only.

It also controls the reconciliation process by running the reconcilers considering the execution dependencies between them.
This reconciler should be run at least every 20 minutes.

#### IstioInstallationReconciler
This reconciler decides if an installation, upgrade or uninstall of Istio in the cluster must be done. The reconciler also creates the IstioOperator
which is used to make changes to the Istio installation by passing it to the `IstioManager`.

The installed IstioOperator is created by merging the `IstioCR` with the IstioOperator with Kyma default values.

##### IstioManager
This component contains the logic for managing the Istio installation. It knows about the supported client versions and forwards the 
Istio API requests (e.g. Install, Upgrade) to the correct version of `IstioClient` if there are breaking changes.  
Depending on how the Istio Go module behaves, it may not be necessary to support multiple clients for the time being. For the first simple implementation, we can probably do without the IstioManager for the time being.

##### IstioClient
A IstioClient encapsulates a specific version of the [Istio Go module](https://github.com/istio/istio). 

> As we want to support canary updates at some point we might need to support two version of the library if there are breaking changes.

Author: We added the support for multiple versions because of one disadvantage mentioned in the [evaluation of Kyma Istio Operator options](https://github.tools.sap/xf-goat/kyma-istio-operator). There it is mentioned that there might be [breaking changes to an interface](https://github.tools.sap/xf-goat/kyma-istio-operator/blob/ec0f99786408407b4a6d8b79abe3af6c389cd35d/README.md?plain=1#L22). 
This alone didn't influence design, but there is some uncertainty how the Istio Go module will evolve. 
If the assumption is correct that installing Istio in version 1.13 also works for uninstalling with it with the version 1.14, we do not need to support two versions as mentioned in the [design](https://github.com/triffer/istio/blob/745581abace17663a730efe7f801f12325cb2ed9/docs/technical-design.md?plain=1#L98). But we need to consider that using the API directly is no officially supported way of installing Istio. There might be some implementations in the supported installation scenarios that take care of this. 
Again, this is really hypothetical and we preferred to draw a solution for this, even if hopefully we will never need it. 
The implementation should be modular enough, so we can easily add support for multiple versions if we have the requirement.

#### ProxySidecarReconciler

> This reconciler must be executed after the `IstioInstallationReconciler`. Its responsibility is to restart pods based on specific configuration changes.

Author: I think this might be partly answered by my reply [here](https://github.com/kyma-project/istio/pull/36#discussion_r1042104795) as we don't really want to have multiple controllers for simplicity.

This means that we do not only want to reconcile when [Istio CR](https://github.com/kyma-project/istio/blob/main/docs/xff-proposal.md) changes, but also want to check time-based that resources have not been changed and are in the expected state.  
The default reconciliation frequency of a manager is defined by the [SyncPeriod](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Options) and is set to 10 hours by default.
So one option is to change the `SyncPeriod` to match the desired reconciliation interval.
Another option is to always return `RequeueAfter` in the result of the Reconcile function to trigger the next reconciliation:
```go
func Reconcile(ctx context.Context, o reconcile.Request) (reconcile.Result, error) {
	// Implement business logic of reading and writing objects here
	return reconcile.Result{RequeueAfter: : 5 * time.Minute}, nil
}
```

The biggest challenge in deciding on an appropriate interval is that the time required to perform the reconciliation can vary a lot. Small changes may only require a 
restart of the sidecar proxies or the Ingress gateway and are therefore much faster than a new installation or a Canary upgrade.

> We can start with using `SyncPeriod` set to 5 minutes as we only want to have a single controller for now and if it's no problem to start with a higher time-based reconciliation. When we add more controllers we can use `RequeueAfter` if

Reviewer 5: ```suggestion
We can start with using `SyncPeriod` set to 5 minutes. For now, we only want to have a single controller, so it is not a problem to start with a higher time-based reconciliation. When we add more controllers we can use `RequeueAfter`
```

> they should be triggered on different intervals or if we want to have a different interval while a long running-process like installation is executed.

Reviewer 5: ```suggestion
in order to trigger them at different intervals, or to have a different interval while a long-running process like installation is executed.
```
